klib/syslog.c

```diff
@@ -0,0 +1,338 @@
+            apply(syslog.fs_write, sg, irangel(syslog.file_offset, len), sh);
+        } else {
+            /* Discard logged data. */
+            deallocate(syslog.h, syslog.file_sgb->buf, SYSLOG_BUF_LEN);
+            kfuncs.deallocate_sg_list(sg);
+        }
+    }
+    syslog_unlock();
+}
+
+define_closure_function(0, 1, void, syslog_timer_func,
+                        u64, overruns)
+{
+    syslog.file_flush_timer = INVALID_ADDRESS;
+    syslog_file_flush();
+}
+
+static void syslog_file_write(void *d, const char *s, bytes count)
+{
+    if (!syslog.fs_write || (count > SYSLOG_BUF_LEN))
+        return;
+    if (!syslog.file_sg) {
+        syslog.file_sg = kfuncs.allocate_sg_list();
+        if (syslog.file_sg == INVALID_ADDRESS)
+            return;
+        syslog.file_sgb = kfuncs.sg_list_tail_add(syslog.file_sg, SYSLOG_BUF_LEN);
+        if (!syslog.file_sgb) {
+            kfuncs.deallocate_sg_list(syslog.file_sg);
+            syslog.file_sg = 0;
+            return;
+        }
+        syslog.file_sgb->buf = allocate(syslog.h, SYSLOG_BUF_LEN);
+        if (syslog.file_sgb->buf == INVALID_ADDRESS) {
+            kfuncs.deallocate_sg_list(syslog.file_sg);
+            syslog.file_sg = 0;
+            return;
+        }
+        syslog.file_sgb->size = SYSLOG_BUF_LEN;
+        syslog.file_sgb->offset = 0;
+        syslog.file_sgb->refcount = 0;
+    }
+    if (syslog.file_sgb->offset + count <= SYSLOG_BUF_LEN) {
+        kfuncs.runtime_memcpy(syslog.file_sgb->buf + syslog.file_sgb->offset, s, count);
+        syslog.file_sgb->offset += count;
+        if (syslog.file_flush_timer == INVALID_ADDRESS) {
+            syslog.file_flush_timer = kfuncs.register_timer(CLOCK_ID_MONOTONIC,
+                SYSLOG_FLUSH_INTERVAL, false, 0, (timer_handler)&syslog.file_flush.__apply);
+        }
+    } else {
+        syslog_file_flush();
+        syslog_file_write(d, s, count);
+    }
+}
+
+static void syslog_dns_cb(const char *name, const ip_addr_t *ipaddr, void *callback_arg)
+{
+    if (ipaddr)
+        syslog.server_ip = *ipaddr;
+    syslog.dns_in_progress = false;
+}
+
+static void syslog_server_resolve(void)
+{
+    err_t err = kfuncs.dns_gethostbyname(syslog.server, &syslog.server_ip, syslog_dns_cb, 0);
+    switch (err) {
+    case ERR_INPROGRESS:
+        syslog.dns_in_progress = true;
+        break;
+    }
+}
+
+static void syslog_set_hdr_len(void)
+{
+    struct netif *n = kfuncs.netif_get_default();
+    if (!n)
+        return;
+    kfuncs.ipaddr_ntoa_r(&n->ip_addr, syslog.local_ip, sizeof(syslog.local_ip));
+    syslog.hdr_len = 1 + sizeof(__XSTRING(SYSLOG_PRIORITY)) + 2 +
+            sizeof("YYYY-MM-ddThh:mm:ss.uuuuuuZ") + runtime_strlen(syslog.local_ip) + 1 +
+            buffer_length(syslog.program) + 5;
+}
+
+static void syslog_udp_write(void *d, const char *s, bytes count)
+{
+    if (ip_addr_isany_val(syslog.server_ip)) {
+        if (!syslog.dns_in_progress)
+            syslog_server_resolve();
```

Comment:
I think some handling of DNS resolution failures would be in order here. Or, at least, some sort of exponential backoff on initiating requests, as it seems that an immediate failure / completion on a lookup could result in possibly many more requests being initiated if there is significant console output.

Reply:
OK, I added an exponential backoff to handle DNS failures.